### PR TITLE
Geomap: Support ability to color markers based on label value

### LINKED ISSUE #47399: Geomap: Support ability to color Geomap markers based on label value

### PR DESCRIPTION
**Pull request by adela-almasan**
https://user-images.githubusercontent.com/88068998/162471719-337eee25-1400-446c-96b8-8da19a8287be.mov

Fixes: #47399 

TODO:
- [x] Fix default option for Mode - updated in #47590

### CODE REVIEW

File: public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts [hunk 1]
```diff
@@ -109,9 +123,22 @@ export const geojsonLayer: MapLayerRegistryItem<GeoJSONMapperConfig> = {
...
+    const style = await getStyleConfigState(config.style);
+    const idToIdx = new Map<string, number>();
+
     const vectorLayer = new VectorLayer({
       source,
       style: (feature: FeatureLike) => {
+        const idx = idToIdx.get(feature.getId() as string);
+        const dims = style.dims;
+
+        if (idx != null && dims !== undefined) {
```
Comment (nmarrs) [inline public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts:135]: Nit
```suggestion
        if (idx && dims) {
```

should check the same conditions unless we specifically care about `null` vs `undefined`

File: public/app/plugins/panel/geomap/layers/data/markersLayer.tsx [hunk 2]
```diff
@@ -119,22 +119,7 @@ export const markersLayer: MapLayerRegistryItem<MarkersConfig> = {
...
-            }
-            if (style.fields.text) {
-              dims.text = getTextDimension(frame, style.config.text!);
-            }
-            if (style.fields.rotation) {
-              dims.rotation = getScalarDimension(frame, style.config.rotation ?? defaultStyleConfig.rotation);
-            }
-            style.dims = dims;
-          }
+          style.dims = getStyleDimension(frame, style, theme);
```
Comment (nmarrs) [inline public/app/plugins/panel/geomap/layers/data/markersLayer.tsx:122]: Great job breaking out this logic into a separate helper function!

**Review by nmarrs — COMMENTED** [review]
Overall looks great! Noticed same small issue of default mode not being set as expected, but once that's cleared up I'm good to approve! I'd be happy to pair on the default mode issue, just let me know :)

**Comment by ryantxu** [comment]
@adela-almasan -- I updated this branch so it adds a *new* layer type rather than trying to split the behavior on the current one.  When https://github.com/grafana/grafana/pull/52760 is ready, I think we can merge this

**Review by ryantxu — APPROVED** [review]
@adela-almasan can you verify and then merge?  this will add a new alpha layer type

**Comment by MatinF** [comment]
This does not seem to work for coords?

I'm trying the below table:

image

As evident, the table includes two different string values for DeviceID. I would expect this to result in two different colors being used in the plot - but all observations have the same color. If I instead of a string use e.g. an integer value, then the color coding works as expected, at least when using threshold palettes. 

image

**Pull request merged**